Fix boolean tensor for map

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108289

torch.empty_strided is able to create a new tensor based on the meta data. For boolean tensor, we call a clone directly, however, we'll get a functional tensor if input is a functional tensor and that functional tensor won't be tracked by tracer's tensor_tracker after dispatching so it become a tensor\_constant in the graph if create_arg. So we manually unwrap the functional tensor before calling clone.


Test Plan:
See added test.

ydwu4 committed:
Fix boolean tensor for map

[ghstack-poisoned]

angelayi reviewed (approved):
I understand the summary, unsure if the implementation is correct 😅 , but the test case makes sense

ydwu4 committed:
Update on "Fix boolean tensor for map"


torch.empty_strided is able to create a new tensor based on the meta data. For boolean tensor, we call a clone directly, however, we'll get a functional tensor if input is a functional tensor and that functional tensor won't be tracked by tracer's tensor_tracker after dispatching so it become a tensor\_constant in the graph if create_arg. So we manually unwrap the functional tensor before calling clone.


Test Plan:
See added test.



[ghstack-poisoned]

ydwu4 committed:
Update on "Fix boolean tensor for map"


torch.empty_strided is able to create a new tensor based on the meta data. For boolean tensor, we call a clone directly, however, we'll get a functional tensor if input is a functional tensor and that functional tensor won't be tracked by tracer's tensor_tracker after dispatching so it become a tensor\_constant in the graph if create_arg. So we manually unwrap the functional tensor before calling clone.


Test Plan:
See added test.



[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
